tools/pubsys/src/aws/client.rs

```diff
@@ -1,159 +1,151 @@
-use rusoto_credential::{
-    AutoRefreshingProvider, AwsCredentials, CredentialsError, DefaultCredentialsProvider,
-    ProfileProvider, ProvideAwsCredentials,
-};
-use rusoto_ebs::EbsClient;
-use rusoto_ec2::Ec2Client;
-use rusoto_ssm::SsmClient;
-use rusoto_sts::{StsAssumeRoleSessionCredentialsProvider, StsClient};
-use snafu::ResultExt;
-
-pub(crate) trait NewWith {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static;
-}
+use aws_config::profile::ProfileFileCredentialsProvider;
 
-impl NewWith for EbsClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+use aws_config::{ConfigLoader, SdkConfig};
+use aws_sdk_ec2::{Endpoint, Region};
+use aws_sdk_sts::model::Credentials as StsCredentials;
+use aws_sdk_sts::{Client as StsClient, Credentials as SdkCredentials};
+use pubsys_config::AwsConfig;
+use snafu::{OptionExt, ResultExt};
+use std::convert::TryFrom;
+use std::time::SystemTime;
 
-impl NewWith for Ec2Client {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+const PROVIDER_NAME: &str = "pubsys";
 
-impl NewWith for SsmClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+pub(crate) async fn build_client_config<S1, S2>(
+    region: S1,
+    sts_region: S2,
+    aws: &AwsConfig,
+) -> Result<SdkConfig>
+where
+    S1: AsRef<str>,
+    S2: AsRef<str>,
+{
+    let mut config = aws_config::from_env();
+    let region = region.as_ref().to_owned();
+    let sts_region = sts_region.as_ref().to_owned();
 
-impl NewWith for StsClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+    // Add custom endpoint if specified in aws.region.REGION.endpoint.
+    let maybe_regional_endpoint = aws.region.get(&region).and_then(|r| r.endpoint.clone());
+    config = if let Some(endpoint) = &maybe_regional_endpoint {
```

Comment:
You probably don't need the additional `config =` assignment here or the `else` (unless I'm missing something).

Since `config` is mutable, you could do:
```
if let Some(...) = ... {
    config.endpoint_resolver...
}
```

Reply:
If `config.endpoint_resolver` isn't a "setter" I suppose my suggestion won't work.

Reply:
Ah - according to the docs it consumes `self` and overrides the endpoint resolver. https://docs.rs/aws-config/latest/aws_config/struct.ConfigLoader.html#method.endpoint_resolver